yarn-project/noir-protocol-circuits/src/crates/rollup-lib/src/base/base_rollup_inputs.nr

```diff
@@ -41,8 +58,373 @@ struct BaseRollupInputs {
+            start_nullifier_tree_snapshot : self.start_nullifier_tree_snapshot,
+            end_nullifier_tree_snapshot : end_nullifier_tree_snapshot,
+            start_contract_tree_snapshot : self.start_contract_tree_snapshot,
+            end_contract_tree_snapshot : end_contract_tree_snapshot,
+            start_public_data_tree_root : self.start_public_data_tree_root,
+            end_public_data_tree_root : end_public_data_tree_root,
+            calldata_hash : calldata_hash,
+        }
+    }
+
+    fn calculate_contract_leaves(self) -> [Field; NUM_CONTRACT_LEAVES] {
+        let mut contract_leaves = [0; NUM_CONTRACT_LEAVES];
+        for i in 0..2 {
+            let new_contracts = self.kernel_data[i].public_inputs.end.new_contracts;
+
+            // loop over the new contracts
+            // TODO(Madiaa): NOTE: we are currently assuming that there is only going to be one
+            for i in 0..new_contracts.len() {
+                let leaf_preimage = new_contracts[i];
+                // When there is no contract deployment, we should insert a zero leaf into the tree and ignore the
+                // member-ship check. This is to ensure that we don't hit "already deployed" errors when we are not
+                // deploying contracts. e.g., when we are only calling functions on existing contracts.
+                let to_push = if leaf_preimage.contract_address.to_field() == 0  {
+                    0
+                } else {    
+                    leaf_preimage.hash()
+                };
+
+                contract_leaves[i] = to_push;
+            }
+        }
+
+        contract_leaves
+    }
+
+    // TODO(Kev): This should say calculate_contract_subtree_root
+    // Cpp code says calculate_contract_subtree, so I'm leaving it as is for now
+    fn calculate_contract_subtree(self, leaves : [Field; NUM_CONTRACT_LEAVES]) -> Field {
+        calculate_subtree(leaves)
+    }
+
+    // TODO(Kev): This should say calculate_commitments_subtree_root
+    // Cpp code says calculate_commitments_subtree, so I'm leaving it as is for now
+    fn calculate_commitments_subtree(self) -> Field {
+        let mut commitment_tree_leaves = [0; NOTE_HASH_SUBTREE_WIDTH];
+        
+        for i in 0..2 {
+            let new_commitments = self.kernel_data[i].public_inputs.end.new_commitments;
+
+            // Our commitments size MUST be 4 to calculate our subtrees correctly
+            assert(new_commitments.len() == MAX_NEW_COMMITMENTS_PER_TX, "New commitments in kernel data must be MAX_NEW_COMMITMENTS_PER_TX (see constants.hpp)");
+
+            for j in 0..MAX_NEW_COMMITMENTS_PER_TX {
+                // TODO(Maddiaa): batch insert
+                commitment_tree_leaves[i * MAX_NEW_COMMITMENTS_PER_TX + j] =  new_commitments[j];
+            }
+        }
+
+        calculate_subtree(commitment_tree_leaves)
+    }
+
+    fn check_nullifier_tree_non_membership_and_insert_to_tree(self) -> AppendOnlyTreeSnapshot {
```

Comment:
This is the main method that is left to implement, ideally we pull out that logic into a separate module named "indexed_merkle_tree" so it can be tested properly but probably okay to do what the cpp code is doing